### PR TITLE
Fix incorrect setting of the SameSite cookie param

### LINKED ISSUE #26377: SameSite cookie, posible issues
| QPT Data Collection|borderStyle=dashed|borderColor=#cccccc|titleBGColor=#dddddd|bgColor=#e3ffea |
| --- |
| 
* **QPT release:**
 1.0.20
* **Patch ID**
MC-41359
* **Are there any additional actions required after the patch installation to make it work?**
 N/A
* **Compatible with Magento versions:**
[Check compatibility](https://web72.us-west-2.stg.sparta.ceng.magento.com/support/patchchk/?patch_id=MC-41359)
|

### Preconditions (*)
On February, 4, Google is set to roll out a new Chrome update that promises a bunch of new features designed to make the browser faster and more secure — including a new approach to cookies.

The SameSite update will require website owners to explicitly state label the third-party cookies that can be used on other sites. Cookies without the proper labelling won’t work in the Chrome browser, which has 63.62% of the overall browser market, according to [Statcounter](https://gs.statcounter.com/browser-market-share).

Right now, the Chrome SameSite cookie default is: “None,” which allows third-party cookies to track users across sites. But from February, cookies will default into “SameSite=Lax,” which means cookies are only set when the domain in the URL of the browser matches the domain of the cookie — a first-party cookie.

This will not probably affect Magento itself but what about it 3rd party integrations which comes pre installed by default such as NewRelic?

### Steps to reproduce (*)
1. Open Chrome and go to [chrome://flags/](chrome://flags/)
1. Enable `SameSite by default cookies` and `Cookies without SameSite must be secure`
1. Open the Chrome inspector. 

### Expected result (*)

1. No errors or warnings should show.

### Actual result (*)

Production site
Screenshot 2020-01-13 at 20 21 44

Admin Panel of a Vanilla Magento 2.3-develop site
Screenshot 2020-01-13 at 20 35 08

Paying with PayPal Express sandbox account
Screenshot 2020-01-13 at 20 46 35

### Related links
- https://www.chromestatus.com/feature/5088147346030592
- https://www.troyhunt.com/promiscuous-cookies-and-their-impending-death-via-the-samesite-policy/
- https://tools.ietf.org/html/draft-west-first-party-cookies-07
- https://www.troyhunt.com/promiscuous-cookies-and-their-impending-death-via-the-samesite-policy/
- https://medium.com/adobetech/adobe-experience-cloud-cookie-updates-for-google-chrome-19ad67cf1598
- https://www.netsparker.com/blog/web-security/same-site-cookie-attribute-prevent-cross-site-request-forgery/
- https://blog.chromium.org/2019/10/developers-get-ready-for-new.html
- https://help.salesforce.com/articleView?id=000351874&language=en_US&type=1&mode=1

### LINKED ISSUE #32440: cookie lax append issue magento 2.4.2 upgrade enterprise b2b
on magento 2.4.2 upgrade Uncaught TypeError: Cannot create property 'cart' on string '{"messages":1613612384,"company":1613612384,"customer":1613612384,"compare-products":1613612384,"requisition":1613612384,"last-ordered-items":1613612384,"cart":1613612384,"directory-data":1613612384,"captcha":1613612384,"wishlist":1613612384,"company_authorization":1613612384,"negotiable_quote":1613612384,"instant-purchase":1613612384,"loggedAsCustomer":1613612384,"multiplewishlist":1613612384,"purchase_order":1613612384,"persistent":1613612384,"review":1613612384,"recently_viewed_product":1613612384,"recently_compared_product":1613612384,"product_data_storage":1613612384,"paypal-billing-agreement":1613612384}laxlaxlaxlaxlaxlax'

### PR DESCRIPTION
**Pull request by ihor-sviziev**
### Description (*)
- Fix incorrect concatenation for the cookie params that lead to appending the `lex` suffix to the value, domain, or other params.
- Add support of the samesite param in places that didn't set it at all

/CC  @hostep 

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/26377
2. Fixes https://github.com/magento/magento2/issues/32440
3. Replaces https://github.com/magento/magento2/pull/32266

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/web/jquery/jquery.cookie.js [hunk 1]
```diff
@@ -47,7 +47,7 @@
                 options.path    ? '; path=' + options.path : '',
                 options.domain  ? '; domain=' + options.domain : '',
                 options.secure  ? '; secure' : '',
-                options.samesite  ? '; samesite=' + options.samesite : 'lax',
+                '; samesite=' + (options.samesite ? options.samesite : 'lax'),
```
Comment (andrewbess) [inline lib/web/jquery/jquery.cookie.js:50]: Hello @ihor-sviziev.
Great changes.
But I have a question.
I am not sure we can change the third-party libraries.
A couple years ago we shouldn't do it.

**Review by andrewbess — CHANGES_REQUESTED** [review]

File: lib/web/jquery/jquery.cookie.js (hunk 1 above)
Comment (ihor-sviziev) [reply lib/web/jquery/jquery.cookie.js:50]: @andrewbess  Actually they were changed in scope of https://github.com/magento/magento2/commit/50e7b94e4fc181ae42150ae52b30e608cba7e426#diff-d0542fed99e2456d704d0d4b4fed14c9c2dedc98b9fb02a3fefc1c229c4f9e14
I'm just fixing these changes

Comment (andrewbess) [reply lib/web/jquery/jquery.cookie.js:50]: Ah, ok.
Thank you for your answer @ihor-sviziev.
Waiting for finishing tests.

File: app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js [hunk 2]
```diff
@@ -17,14 +17,17 @@ define(function () {
         function setFormKeyCookie(value) {
             var expires,
                 secure,
+                samesite,
                 date = new Date(),
-                isSecure = !!window.cookiesConfig && window.cookiesConfig.secure;
+                cookiesConfig = window.cookiesConfig || {},
+                isSecure = !!cookiesConfig.secure;
```
Comment (Den4ik) [inline app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js]:
```suggestion
                cookiesConfig = window.cookiesConfig || {},
                isSecure = !!cookiesConfig.secure;
                samesite = cookiesConfig.samesite || 'lax';
```

File: app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js [hunk 3]
```diff
@@ -17,14 +17,17 @@ define(function () {
         function setFormKeyCookie(value) {
             var expires,
                 secure,
+                samesite,
```
Comment (Den4ik) [inline app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js]:
```suggestion
```

File: app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js [hunk 4]
```diff
@@ -17,14 +17,17 @@ define(function () {
...
                 date = new Date(),
-                isSecure = !!window.cookiesConfig && window.cookiesConfig.secure;
+                cookiesConfig = window.cookiesConfig || {},
+                isSecure = !!cookiesConfig.secure;
 
             date.setTime(date.getTime() + 86400000);
             expires = '; expires=' + date.toUTCString();
             secure = isSecure ? '; secure' : '';
+            samesite = '; samesite=' + (cookiesConfig.samesite ? cookiesConfig.samesite : 'lax');
```
Comment (Den4ik) [inline app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js]:
```suggestion
            samesite = '; samesite=' + samesite;
```

File: lib/web/mage/adminhtml/tools.js [hunk 5]
```diff
@@ -267,7 +267,7 @@ var Cookie = {
 
         return null;
     },
-    write: function (cookieName, cookieValue, cookieLifeTime) {
+    write: function (cookieName, cookieValue, cookieLifeTime, samesite) {
```
Comment (Den4ik) [inline lib/web/mage/adminhtml/tools.js:270]:
```suggestion
    write: function (cookieName, cookieValue, cookieLifeTime, samesite = 'lax') {
```

File: lib/web/mage/adminhtml/tools.js [hunk 6]
```diff
@@ -278,7 +278,9 @@ var Cookie = {
         }
         var urlPath = '/' + BASE_URL.split('/').slice(3).join('/'); // Get relative path
 
-        document.cookie = escape(cookieName) + '=' + escape(cookieValue) + expires + '; path=' + urlPath;
+        samesite = '; samesite=' + (samesite ? samesite : 'lax');
```
Comment (Den4ik) [inline lib/web/mage/adminhtml/tools.js:281]:
```suggestion
        samesite = '; samesite=' + samesite;
```

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi, @ihor-sviziev. Thanks for the fixes.
Please, look at my suggestions, that doesn't change the main logic implemented by your PR, but improve code style

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js [hunk 7]
```diff
@@ -18,13 +18,16 @@ define(function () {
             var expires,
                 secure,
                 date = new Date(),
-                isSecure = !!window.cookiesConfig && window.cookiesConfig.secure;
+                cookiesConfig = window.cookiesConfig || {},
+                isSecure = !!cookiesConfig.secure;
```
Comment (Den4ik) [inline app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js]:
```suggestion
                isSecure = !!cookiesConfig.secure,
```

Comment (Den4ik) [reply app/code/Magento/PageCache/view/frontend/web/js/form-key-provider.js]: @ihor-sviziev my fail with the previous suggestion

**Review by Den4ik — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Review by andrewbess — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @andrewbess, thank you for the review.
[ENGCOM-8927](https://jira.corp.magento.com/browse/ENGCOM-8927) has been created to process this Pull Request
:eight_spoked_asterisk: @andrewbess, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by Den4ik — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-8927](https://jira.corp.magento.com/browse/ENGCOM-8927) has been created to process this Pull Request
:eight_spoked_asterisk: @Den4ik, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Integration Tests

**Comment by engcom-Oscar** [comment]
@engcom-Foxtrot can you please check this PR, it require dev-experience.

**Comment by sivamind** [comment]
is there any immediate patch is available? how to include these changes in magento commerce cloud project

**Review by engcom-Foxtrot — CHANGES_REQUESTED** [review]
Hello, @ihor-sviziev. Thank you for your contribution. Unfortunately, we were unable to reproduce an issue you are fixing on this PR. Could you please add some steps to reproduce? Thanks in advance.

**Comment by ihor-sviziev** [comment]
@magnento give me 2.4-develop instance

**Comment by ihor-sviziev** [comment]
@magento give me 2.4-develop instance

**Review by engcom-Foxtrot — APPROVED** [review]
QA passed

**Comment by ihor-sviziev** [comment]
I'll just add some info about reproducing this issue that we discussed with @engcom-Foxtrot in Slack:

**Example 1:**

1. Go to frontend, open developer tools
2. Add breakpoint before document.cookie in the mage/cookie.js file.
3. Run in console following code:
```
jQuery.mage.cookies.set('test', 'value', {samesite: null})
```
4. See what going to be set in `document.cookie`

**Actual result** (before this PR)
❌ added 'lax' to the end of cookie domain; no samesite attribute
image

**Expected result** (with this PR)
✔ "lax" shouldn't be added to the cookie domain; the samesite attribute should be present with default value - "lax"


**Example 2:**

1. Go to frontend, open developer tools
2. Add breakpoint before document.cookie in the jquery.cookie.js file.
3. Run in console following code:
```
jQuery.cookie('foo', 'bar', {domain: 'foobar.com', samesite: null})
```
4. See what going to be set in `document.cookie`

**Actual result** (before this PR)
❌ added 'lax' to the end of cookie domain; no samesite attribute
the output looks like this:
```
"foo=bar; path=/; domain=foobar.comlax"
```

**Expected result** (with this PR)
✔ "lax" shouldn't be added to the cookie domain; the samesite attribute should be present with default value - "lax"

**Comment by mrtuvn** [comment]
Should we change capitalize first letter of 'lax' ? Change to 'Lax'

**Comment by ihor-sviziev** [comment]
@mrtuvn according to RFC - it's case insensitive:
https://tools.ietf.org/id/draft-ietf-httpbis-rfc6265bis-03.html#the-samesite-attribute-1

**Comment by zaietsv** [comment]
While waiting for a regular fix I can propose an easy way to workaround a "}lax" problem in cookies. As for me I don't like to penetrate deep into such complicated materials like overriding core jquery libraries. So I suggest to override a phtml template.

For a hot-fix it is necessary to customize at your theme either at any other module a template **vendor/magento/module-cookie/view/base/templates/html/cookie.phtml** .  You need to add to the template a new row
  `window.cookiesConfig.samesite = 'lax';`
 which initializes a **"samesite"** value at a storefront.

[http://screenshots.collabstar.com/vza/Selection_d75f109.png](url)

This will initialize a previously undefined field **"_samesite"** at a frontend script  **vendor/magento/module-cookie/view/base/web/js/jquery.storageapi.extended.js**
`_samesite: window.cookiesConfig ? window.cookiesConfig.samesite : 'lax',`

[http://screenshots.collabstar.com/vza/Selection_53b8cb2.png](url)

The above mentioned operation allows to bypass a buggy part of a statement 
`options.samesite  ? '; samesite=' + options.samesite : 'lax'`
at a script **lib/web/jquery/jquery.cookie.js** .
[http://screenshots.collabstar.com/vza/Selection_255d05c.png](url)

Since a "false" branch of the statement is incorrect and a "true" branch is OK we only need to define an argument **"options.samesite"**. This allows to implement a positive scenario and omit a wrong "lax" assignment.

Of course one can implement the fix in a gentle way by providing additional arguments to a block 
`block class="Magento\Framework\View\Element\Js\Cookie" name="cookie_config"` 
at a layout. 

Have a nice fixing!

**Comment by ihor-sviziev** [comment]
Just note the fix also available in the quality patches tool as MC-41359.

**Pull request merged**